README
NAME

    Clustericious::Config - Configuration files for Clustericious nodes.

VERSION

    version 0.30

SYNOPSIS

    In your ~/etc/MyApp.conf file:

     ---
     % extends_config 'global';
     % extends_config 'hypnotoad', url => 'http://localhost:9999', app => 'MyApp';
    
     url : http://localhost:9999
     start_mode : hypnotoad
     hypnotoad :
        - heartbeat_timeout : 500
     
     arbitrary_key: value

    In your ~/etc/globa.conf file:

     ---
     somevar : somevalue

    In your ~/etc/hypnotoad.conf:

     listen :
        - <%= $url %>
     # home uses File::HomeDir to find the calling users'
     # home directory
     pid_file : <%= home %>/<%= $app %>/hypnotoad.pid
     env :
        MOJO_HOME : <%= home %>/<%= $app %>

    From a Clustericious::App:

     package MyApp;
     
     use Mojo::Base qw( Clustericious::App );
     
     package MyApp::Routes;
     
     use Clustericious::RouteBuilder;
     
     get '/' => sub {
       my $c = shift;
       my $config = $c; # $config isa Clustericious::Config
       
       # returns the value if it is defined, foo otherwise
       my $value = $config->arbitrary_key(default => 'foo');
     };

    From a script:

     use Clustericious::Config;
     
     my $c = Clustericious::Config->new("MyApp");
     my $c = Clustericious::Config->new( \$config_string );
     my $c = Clustericious::Config->new( \%config_data_structure );
    
     print $c->url;
     print $c->{url};
    
     print $c->hypnotoad->listen;
     print $c->hypnotoad->{listen};
     my %hash = $c->hypnotoad;
     my @ary  = $c->hypnotoad;
    
     # Supply a default value for a missing configuration parameter :
     $c->url(default => "http://localhost:9999");
     print $c->this_param_is_missing(default => "something_else");
    
     # Dump out the entire config as yaml
     print $c->dump_as_yaml;

DESCRIPTION

    Clustericious::Config reads configuration files which are
    Mojo::Template's of JSON or YAML files. There should generally be an
    entry for 'url', which may be used by either a client or a server
    depending on how this node in the cluster is being used.

    After rendering the template and parsing the JSON, the resulting object
    may be called using method calls or treated as hashes.

    Config files are looked for in the following places (in order, where
    "MyApp" is the name of the app) :

        $CLUSTERICIOUS_CONF_DIR/MyApp.conf
        $HOME/etc/MyApp.conf
        /util/etc/MyApp.conf
        /etc/MyApp.conf

    The helper "extends_config" may be used to read default settings from
    another config file. The first argument to extends_config is the
    basename of the config file. Additional named arguments may be passed
    to that config file and used as variables within that file. After
    reading another file, the hashes are merged (i.e. with Hash::Merge); so
    values anywhere inside the data structure may be overridden.

    YAML config files must begin with "---", otherwise they are interpreted
    as JSON.

    This module provides a number of helpers which can be used to get
    system details (such as the home directory of the calling user or to
    prompt for passwords). See Clustericious::Config::Helpers for details.

CONSTRUCTOR

 new

    Create a new Clustericious::Config object. See the SYNOPSIS for
    possible invocations.

METHODS

 $config->dump_as_yaml

    Returns a string with the configuration encoded as YAML.

 Clustericious::Config->set_singleton

    Cache a config object to be returned by the constructor. Usage:

     Clustericicious::Config->set_singleton(App => $object);

CAVEATS

    Some filesystems do not support filenames with a colon (:) character in
    them, so for apps with a double colon in them (for example
    Clustericious::HelloWorld), a single dash character will be substituted
    for the name (for example Clustericious-HelloWorld.conf).

NOTES

    This is a beta release. The API may change without notice.

SEE ALSO

    Mojo::Template, Hash::Merge, Clustericious, Clustericious::Client,
    Clustericious::Config::Helpers

AUTHOR

    original author: Brian Duggan

    current maintainer: Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2013 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Clustericious-Config

0.30      2014-10-07 05:45:25 America/New_York
  - fix test error for when File::HomeDir->users_home is not implemented

0.29      2014-09-09 16:29:06 America/New_York
  - added yaml helper for dumping yaml

0.28      2014-04-09 12:06:48 America/New_York
  - remove test harness detection

0.27      2013-11-26 11:23:26 America/New_York
  - rm dep on IO::File

0.26      2013-08-31 14:31:30 America/New_York
  - require Path::Class 0.26 for spew()

0.25      2013-07-18 09:20:51 America/New_York
  - promote to production

0.24_06   2013-07-17 15:48:26 America/New_York
  - rename Clustericious::Config::Plugin to ::Helpers
  - create a deprecated module ::Plugin to keep Test::Clustericious::Cluster working

0.24_05   2013-07-16 05:54:20 America/New_York
  - migrate to Dist::Zilla

0.24_04 Sun Jul 14 23:04:22 EDT 2013
        - added json helper for config files

0.24_03 Thu Jul 11 15:54:31 EDT 2013
        - bug fix: file returns a Path::Class::File instead of ::Dir

0.24_02 Thu Jul 11 15:42:37 EDT 2013
        - added hostname helper for config files
        - added hostname_full helper for config files

0.24_01 Wed Jul 10 22:02:15 EDT 2013
        - converted clconfig into clustericious configtest

0.24    Mon Jul  8 16:40:45 EDT 2013
        - added home helper for config files
        - added file helper for config files
        - added dir helper for config files

0.23    Sun Jul  7 21:27:21 EDT 2013
        - substitute - for :: in config filenames
          for example ~/etc/Clustericious::HelloWorld.conf should now be
          ~/etc/Clustericious-HelloWorld.conf

0.22    Wed Jul  3 15:00:09 EDT 2013
        - added create_config_helper_ok to Test::Clustericious::Config
        - don't re-use File::HomeDir::Test if it has already been loaded

0.21    Sun Jun 30 18:23:49 EDT 2013
        - added example to Test::Clustericious::Config
        - add clconfig debug script

0.20    Wed Jun 26 21:47:02 EDT 2013
        - require File::HomeDir 0.91 (need File::HomeDir::Test)

0.18    Fri Jun 14 10:45:31 EDT 2013
        - put template packages in Config::Clustericious::Config::TemplatePackage::Package{1,2,3...}
        - put the app config package in Clustericious::Config::App::AppName

0.17    Fri Jun 14 10:41:56 EDT 2013
        - added Test::Clustericious::Config

0.16    Wed Jun  5 13:21:54 EDT 2013
        - Documentation

0.15    Wed Jun  5 10:05:10 EDT 2013
        Added back changes from 0.13 that got reverted in 0.14

0.14    Tue Jun  4 07:56:15 EDT 2013
        Run template in different package

0.13    Sat Jun  1 13:23:55 EDT 2013
        fixed autoload conflicts, home, dclone, etc.

0.12    Mon Aug 27 10:38:42 EDT 2012
        portability improvements (Graham)

0.11    Tue Aug 21 17:14:45 EDT 2012
        support Makefile.PL

0.10    Thu Aug 16 11:49:10 EDT 2012
        first release




lib/Test/Clustericious/Config.pm
package Test::Clustericious::Config;

use strict;
use warnings;
use v5.10;

BEGIN {
  unless($INC{'File/HomeDir/Test.pm'})
  {
    eval q{ use File::HomeDir::Test };
    die $@ if $@;
  }
}

use File::HomeDir;
use YAML::XS qw( DumpFile );
use File::Path qw( mkpath );
use Clustericious::Config;
use Mojo::Loader;

use base qw( Test::Builder::Module Exporter );

our @EXPORT = qw( create_config_ok create_directory_ok home_directory_ok create_config_helper_ok );
our @EXPORT_OK = @EXPORT;
our %EXPORT_TAGS = ( all => \@EXPORT );

my $config_dir;

sub _init
{
  $config_dir = File::HomeDir->my_home . "/etc";
  mkdir $config_dir;

  $ENV{CLUSTERICIOUS_CONF_DIR} = $config_dir;
  Clustericious::Config->_testing(1);
}

BEGIN { _init() }

# ABSTRACT: Test Clustericious::Config
our $VERSION = '0.30'; # VERSION


sub create_config_ok ($;$$)
{
  my($config_name, $config, $test_name) = @_;

  my $fn = "$config_name.conf";
  $fn =~ s/::/-/g;
  
  unless(defined $config)
  {
    my $loader = Mojo::Loader->new;
    my $caller = caller;
    $loader->load($caller);
    $config = $loader->data($caller, "etc/$fn");
  }
  
  my $tb = __PACKAGE__->builder;  
  my $ok = 1;
  unless(defined $config)
  {
    $config = "---\n";
    $tb->diag("unable to locate text for $config_name");
    $ok = 0;
  }
  
  my $config_filename = "$config_dir/$fn";
  
  eval {
    if(ref $config)
    {
      DumpFile($config_filename, $config);
    }
    else
    {
      open my $fh, '>', $config_filename;
      print $fh $config;
      close $fh;
    }
  };
  if(my $error = $@)
  {
    $ok = 0;
    $tb->diag("exception: $error");
  }
  
  $test_name //= "create config for $config_name at $config_filename";
  
  # remove any cached copy if necessary
  Clustericious::Config->_uncache($config_name);

  $tb->ok($ok, $test_name);
  return $config_filename;
}


sub create_directory_ok ($;$)
{
  my($path, $test_name) = @_;
  
  my $fullpath = $path;
  $fullpath =~ s{^/}{};
  $fullpath = join('/', File::HomeDir->my_home, $fullpath);
  mkpath $fullpath, 0, 0700;
  
  $test_name //= "create directory $fullpath";
  
  my $tb = __PACKAGE__->builder;
  $tb->ok(-d $fullpath, $test_name);
  return $fullpath;
}


sub home_directory_ok (;$)
{
  my($test_name) = @_;
  
  my $fullpath = File::HomeDir->my_home;
  
  $test_name //= "home directory $fullpath";
  
  my $tb = __PACKAGE__->builder;
  $tb->ok(-d $fullpath, $test_name);
  return $fullpath;
}


sub create_config_helper_ok ($$;$)
{
  my($helper_name, $helper_code, $test_name) = @_;
  
  $test_name //= "create config helper $helper_name";
  
  require Clustericious::Config::Helpers;
  do {
    no strict 'refs';
    *{"Clustericious::Config::Helpers::$helper_name"} = $helper_code;
  };
  push @Clustericious::Config::Helpers::EXPORT, $helper_name;
  
  my $tb = __PACKAGE__->builder;
  $tb->ok(1, $test_name);
  return;
}

1;


__END__
=pod

=head1 NAME

Test::Clustericious::Config - Test Clustericious::Config

=head1 VERSION

version 0.30

=head1 SYNOPSIS

 use Test::Clustericious::Config;
 use Clustericious::Config;
 use Test::More tets => 2;
 
 create_config_ok 'Foo', { url => 'http://localhost:1234' };
 my $config = Clustericious::Config->new('Foo');
 is $config->url, "http://localhost:1234";

To test against a Clustericious application MyApp:

 use Test::Clustericious::Config;
 use Test::Clustericious;
 use Test::More tests => 3;

 create_config_ok 'MyApp', { x => 1, y => 2 }; 
 my $t = Test::Clustericious->new('MyApp');
 
 $t->get_ok('/');
 
 is $t->app->config->x, 1;

To test against multiple Clustericious applications MyApp1, MyApp2
(can also be the same app with different config):

 use Test::Clustericious::Config;
 use Test::Clustericious;
 use Test::More tests => 4;
 
 create_config_ok 'MyApp1', {};
 my $t1 = Test::Clustericious->new('MyApp1');
 
 $t1->get_ok('/');
 
 create_config_ok 'MyApp2', { my_app1_url => $t1->app_url };
 my $t2 = Test::Clustericious->new('MyApp2');
 
 $t2->get_ok('/');

=head1 DESCRIPTION

This module provides an interface for testing Clustericious
configurations, or Clustericious applications which use
a Clustericious configuration.

It uses L<File::HomeDir::Test> to isolate your test environment
from any configurations you may have in your C<~/etc>.  Keep
in mind that this means that C<$HOME> and friends will be in
a temporary directory and removed after the test runs.  It also
means that the caveats for L<File::HomeDir::Test> apply when
using this module as well (ie. this should be the first module
that you use in your test after C<use strict> and C<use warnings>).

=head1 FUNCTIONS

=head2 create_config_ok $name, $config, [$test_name]

Create a Clustericious config with the given C<$name>.
If C<$config> is a reference then it will create the 
configuration file with C<YAML::XS::DumpFile>, if
it is a scalar, it will will write the scalar out
to the config file.  Thus these three examples should
create a config with the same values (though in different
formats):

hash reference:

 create_config_ok 'Foo', { url => 'http://localhost:1234' }];

YAML:

 create_config_ok 'Foo', <<EOF;
 ---
 url: http://localhost:1234
 EOF

JSON:

 create_config_ok 'Foo', <<EOF;
 {"url":"http://localhost:1234"}
 EOF

In addition to being a test that will produce a ok/not ok
result as output, this function will return the full path
to the configuration file created.

=head2 create_directory_ok $path, [$test_name]

Creates a directory in your test environment home directory.
This directory will be recursively removed when your test
terminates.  This function returns the full path of the 
directory created.

=head2 home_directory_ok [$test_name]

Tests that the temp home directory has been created okay.
Returns the full path of the home directory.

=head2 create_config_helper_ok $helper_name, $helper_coderef, [ $test_name ]

Install a helper which can be called from within a configuration template.
Example:

 my $counter;
 create_config_helper_ok 'counter', sub { $counter++ };
 create_config_ok 'MyApp', <<EOF;
 ---
 one: <%= counter %>
 two: <%= counter %>
 three: <% counter %>
 EOF

=head1 EXAMPLES

Here is an (abbreviated) example from L<Yars> that show how to test against an app
where you need to know the port/url of the app in the configuration
file:

 use Test::Mojo;
 use Test::More tests => 1;
 use Test::Clustericious::Config;
 use Mojo::UserAgent;
 use Yars;
 
 my $t = Test::Mojo->new;
 $t->ua(do {
   my $ua = Mojo::UserAgent->new;
   create_config_ok 'Yars', {
     url => $ua->app_url,
     servers => [ {
       url => $ua->app_url,
     } ]
   };
   $ua->app(Yars->new);
   $ua
 };
 
 $t->get_ok('/status');

To see the full tests see t/073_tempdir.t in the L<Yars> distribution.

=head1 AUTHOR

original author: Brian Duggan

current maintainer: Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Configuration files for Clustericious nodes.",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "Clustericious-Config",
  "dist_version" => "0.30",
  "license" => "perl",
  "module_name" => "Clustericious::Config",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Cwd" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "File::HomeDir" => 0,
    "File::Path" => 0,
    "Hash::Merge" => 0,
    "JSON::XS" => 0,
    "List::Util" => 0,
    "Log::Log4perl" => 0,
    "Module::Build" => "0.3601",
    "Mojolicious" => "3.85",
    "Path::Class" => "0.26",
    "Storable" => 0,
    "Sys::Hostname" => 0,
    "Term::Prompt" => 0,
    "Test::Builder::Module" => 0,
    "YAML::XS" => 0,
    "base" => 0,
    "perl" => "v5.10.0"
  },
  "script_files" => [],
  "test_requires" => {
    "File::HomeDir::Test" => 0,
    "File::Temp" => 0,
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "File::HomeDir::Test" => 0,
  "File::Temp" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Clustericious/Config/Plugin.pm
package Clustericious::Config::Plugin;

use strict;
use warnings;
use Clustericious::Config::Helpers;

# ABSTRACT: Deprecated module
our $VERSION = '0.30'; # VERSION


# Hack to keep old versions of Test::Clustericious::Cluster working
*EXPORT = \@Clustericious::Config::Helpers::EXPORT;

sub Clustericious::Config::Helpers::cluster
{
  goto &Clustericious::Config::Plugin::cluster;
}

1;

__END__
=pod

=head1 NAME

Clustericious::Config::Plugin - Deprecated module

=head1 VERSION

version 0.30

=head1 DESCRIPTION

This is module is deprecated, please see
L<Clustericious::Config::Helper>.

=head1 AUTHOR

original author: Brian Duggan

current maintainer: Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Clustericious/Config/Helpers.pm
package Clustericious::Config::Helpers;

use strict;
use warnings;
use v5.10;
use Hash::Merge qw/merge/;
use Data::Dumper;
use Carp qw( croak );
use base qw( Exporter );
use JSON::XS qw( encode_json );

# ABSTRACT: Helpers for clustericious config files.
our $VERSION = '0.30'; # VERSION


our @mergeStack;
our @EXPORT = qw( extends_config get_password home file dir hostname hostname_full json yaml );


sub extends_config {
    my $filename = shift;
    my @args = @_;
    push @mergeStack, Clustericious::Config->new($filename, \@args);
    return '';
}

#
#
# do_merges:
#
# Called after reading all config files, to process extends_config
# directives.
#
sub _do_merges {
    my $class = shift;
    my $conf_data = shift; # Last one; Has highest precedence.

    return $conf_data unless @mergeStack;

    # Nested extends_config's form a tree which we traverse depth first.
    Hash::Merge::set_behavior( 'RIGHT_PRECEDENT' );
    my %so_far = %{ shift @mergeStack };
    while (my $c = shift @mergeStack) {
        my %h = %$c;
        %so_far = %{ merge( \%so_far, \%h ) };
    }
    %$conf_data = %{ merge( \%so_far, $conf_data ) };
}


sub get_password {
    return Clustericious::Config::Password->sentinel;
}


sub home (;$)
{
  require File::HomeDir;
  $_[0] ? File::HomeDir->users_home($_[0]) : File::HomeDir->my_home;
}


sub file
{
  eval { require Path::Class::File };
  croak "file helper requires Path::Class" if $@;
  Path::Class::File->new(@_);
}


sub dir
{
  require Path::Class::Dir;
  croak "dir helper requires Path::Class" if $@;
  Path::Class::Dir->new(@_);
}


sub hostname
{
  state $hostname;
  
  unless(defined $hostname)
  {
    require Sys::Hostname;
    $hostname = Sys::Hostname::hostname();
    $hostname =~ s/\..*$//;
  }
  
  $hostname;
}


sub hostname_full
{
  require Sys::Hostname;
  Sys::Hostname::hostname();
}


sub json ($)
{
  encode_json($_[0]);
}


sub yaml ($)
{
  require YAML::XS;
  local $YAML::UseHeader = 0;
  my $str = YAML::XS::Dump($_[0]);
  $str =~ s{^---\n}{};
  $str;
}


1;

__END__
=pod

=head1 NAME

Clustericious::Config::Helpers - Helpers for clustericious config files.

=head1 VERSION

version 0.30

=head1 SYNOPSIS

 ---
 % extend_config 'SomeOtherConfig';

=head1 DESCRIPTION

This module provides the functions available in all configuration files
using L<Clustericious::Config>.

=head1 FUNCTIONS

=head2 extends_config $config_name, %arguments

Extend the config using another config file.

=head2 get_password

Prompt for a password.  This will prompt the user the first time it is
encountered for a password.

=head2 home( [ $user ] )

Return the given users' home directory, or if no user is
specified return the calling user's home directory.

=head2 file( @list )

The C<file> shortcut from Path::Class, if it is installed.

=head2 dir( @list )

The C<dir> shortcut from Path::Class, if it is installed.

=head2 hostname

The system hostname (uses L<Sys::Hostname>)

=head2 hostname_full

The system hostname in full, including the domain, if
it can be determined (uses L<Sys::Hostname>).

=head2 json $ref

Encode the given hash or list reference.

=head2 yaml $ref

Encode the given hash or list reference.

=head1 SEE ALSO

L<Clustericious::Config>, L<Clustericious>

=head1 AUTHOR

original author: Brian Duggan

current maintainer: Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Clustericious/Config/Password.pm
package Clustericious::Config::Password;

use Data::Dumper;
use strict;
use warnings;

# ABSTRACT: Password routines for Clustericious::Config
our $VERSION = '0.30'; # VERSION


our $Stashed;

sub sentinel {
    return "__XXX_placeholder_ceaa5b9c080d69ccdaef9f81bf792341__";
}

sub get {
    my $self = shift;
    require Term::Prompt;
    $Stashed ||= Term::Prompt::prompt('p', 'Password:', '', '');
    $Stashed;
}

sub is_sentinel {
    my $class = shift;
    my $val = shift;
    return (defined($val) && $val eq $class->sentinel);
}

1;


__END__
=pod

=head1 NAME

Clustericious::Config::Password - Password routines for Clustericious::Config

=head1 VERSION

version 0.30

=head1 DESCRIPTION

This module provides the machinery for handling passwords used by
L<Clustericious::Config> and L<Clustericious::Config::Helpers>.

=head1 AUTHOR

original author: Brian Duggan

current maintainer: Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Clustericious/Config.pm
package Clustericious::Config;

use strict;
use warnings;
use v5.10;

# ABSTRACT: Configuration files for Clustericious nodes.
our $VERSION = '0.30'; # VERSION


use Clustericious::Config::Password;
use List::Util;
use JSON::XS;
use YAML::XS ();
use Mojo::Template;
use Log::Log4perl qw/:easy/;
use Storable;
use Clustericious::Config::Helpers ();
use Data::Dumper;
use Cwd ();
use Module::Build;
use File::HomeDir ();

our %Singletons;

sub _is_subdir {
    my ($child,$parent) = @_;
    my $p = Cwd::abs_path($parent);
    my $c = Cwd::abs_path($child);
    return ($c =~ m[^\Q$p\E]) ? 1 : 0;
}

my $is_test = 0;
sub _testing {
  my($class, $new) = @_;
  $is_test = $new if defined $new;
  $is_test;
}

our $class_suffix = {};
sub _uncache {
    my($class, $name) = @_;
    delete $Clustericious::Config::Singletons{$name};
    $class_suffix->{$name} //= 1;
    $class_suffix->{$name}++;
}

sub pre_rendered { }
sub rendered { }


sub new {
    my $class = shift;
    my %t_args = (ref $_[-1] eq 'ARRAY' ? @{( pop )} : () );
    my $arg = $_[0];
    ($arg = caller) =~ s/:.*$// unless $arg; # Determine from caller's class
    return $Singletons{$arg} if exists($Singletons{$arg});

    
    my $we_are_testing_this_module = 0;
    if(__PACKAGE__->_testing) {
        $we_are_testing_this_module = 0;
    }

    my $conf_data;

    my $json = JSON::XS->new;
    
    state $package_counter = 0;
    my $namespace = "Clustericious::Config::TemplatePackage::Package$package_counter";
    eval qq{ package $namespace; use Clustericious::Config::Helpers; };
    die $@ if $@;
    $package_counter++;
    
    my $mt = Mojo::Template->new(namespace => $namespace)->auto_escape(0);
    $mt->prepend( join "\n", map " my \$$_ = q{$t_args{$_}};", sort keys %t_args );

    my $filename;
    if (ref $arg eq 'SCALAR') {
        $class->pre_rendered( $$arg );
        my $rendered = $mt->render($$arg);
        $class->rendered( SCALAR => $rendered );
        die $rendered if ( (ref($rendered)) =~ /Exception/ );
        my $type = $rendered =~ /^---/ ? 'yaml' : 'json';
        $conf_data = $type eq 'yaml' ?
           eval { YAML::XS::Load( $rendered ); }
         : eval { $json->decode( $rendered ); };
        LOGDIE "Could not parse $type \n-------\n$rendered\n---------\n$@\n" if $@;
    } elsif (ref $arg eq 'HASH') {
        $conf_data = Storable::dclone $arg;
    } elsif (
          $we_are_testing_this_module
          && !(
              $ENV{CLUSTERICIOUS_CONF_DIR}
              && _is_subdir( $ENV{CLUSTERICIOUS_CONF_DIR}, Cwd::getcwd() )
          )) {
          $conf_data = {};
    } else {
        my @conf_dirs;

        @conf_dirs = $ENV{CLUSTERICIOUS_CONF_DIR} if defined( $ENV{CLUSTERICIOUS_CONF_DIR} );

        push @conf_dirs, ( File::HomeDir->my_home . "/etc", "/util/etc", "/etc" ) unless $we_are_testing_this_module || __PACKAGE__->_testing;
        my $conf_file = "$arg.conf";
        $conf_file =~ s/::/-/g;
        my ($dir) = List::Util::first { -e "$_/$conf_file" } @conf_dirs;
        if ($dir) {
            TRACE "reading from config file $dir/$conf_file";
            $filename = "$dir/$conf_file";
            $class->pre_rendered( $filename );
            my $rendered = $mt->render_file($filename);
            $class->rendered( $filename => $rendered );
            die $rendered if ( (ref $rendered) =~ /Exception/ );
            my $type = $rendered =~ /^---/ ? 'yaml' : 'json';
            if ($ENV{CL_CONF_TRACE}) {
                warn "configuration ($type) : \n";
                warn $rendered;
            }
            $conf_data =
              $type eq 'yaml'
              ? eval { YAML::XS::Load($rendered) }
              : eval { $json->decode($rendered) };
            LOGDIE "Could not parse $type\n-------\n$rendered\n---------\n$@\n" if $@;
        } else {
            TRACE "could not find $conf_file file in: @conf_dirs" unless $dir;
            $conf_data = {};
        }
    }
    $conf_data ||= {};
    Clustericious::Config::Helpers->_do_merges($conf_data);
    _add_heuristics($filename,$conf_data);
    # Use derived classes so that AUTOLOADING keeps namespaces separate
    # for various apps.
    if ($class eq __PACKAGE__) {
        if (ref $arg) {
            $arg = "$arg";
            $arg =~ tr/a-zA-Z0-9//cd;
        }
        $class = join '::', $class, 'App', $arg;
        $class .= $class_suffix->{$arg} if $class_suffix->{$arg};
        my $dome = '@'."$class"."::ISA = ('".__PACKAGE__. "')";
        eval $dome;
        die "error setting ISA : $@" if $@;
    }
    bless $conf_data, $class;
}

sub _add_heuristics {
    my $filename = shift;
    # Account for some mojo api changes
    my $conf_data = shift;
    if ($conf_data->{hypnotoad} && !ref($conf_data->{hypnotoad}{listen})) {
        warn "# hypnotoad->listen should be an arrayref in $filename\n";
        $conf_data->{hypnotoad}{listen} = [ $conf_data->{hypnotoad}{listen} ];
    }


}


sub dump_as_yaml {
    my $c = shift;
    return YAML::XS::Dump($c);
}

sub _stringify {
    my $self = shift;
    return join ' ', map { ($_, $self->{$_}) } sort keys %$self;
}

sub DESTROY {
}

sub AUTOLOAD {
    my $self = shift;
    my %args = @_;
    my $default = $args{default};
    my $default_exists = exists $args{default};
    our $AUTOLOAD;
    my $called = $AUTOLOAD;
    $called =~ s/.*:://g;
    if ($default_exists && !exists($self->{$called})) {
        $self->{$called} = $args{default};
    }
    Carp::cluck "config element '$called' not found for ".(ref $self)." (".(join ',',keys(%$self)).")"
        if $called =~ /^_/ || !exists($self->{$called});
    my $value = $self->{$called};
    my $obj;
    my $invocant = ref $self;
    if (ref $value eq 'HASH') {
        $obj = $invocant->new($value);
    }
    no strict 'refs';
    *{ $invocant . "::$called" } = sub {
          my $self = shift;
          $self->{$called} = $default if $default_exists && !exists($self->{$called});
          die "'$called' not found in ".join ',',keys(%$self)
              unless exists($self->{$called});
          my $value = $self->{$called};
          return wantarray && (ref $value eq 'HASH' ) ? %$value
          : wantarray && (ref $value eq 'ARRAY') ? @$value
          :                       defined($obj)  ? $obj
          : Clustericious::Config::Password->is_sentinel($value) ? Clustericious::Config::Password->get
          :                                        $value;
    };
    use strict 'refs';
    $self->$called;
}


sub set_singleton {
    my $class = shift;
    my $app = shift;
    my $obj = shift;
    our %Singletons;
    $Singletons{$app} = $obj;
}


1;


__END__
=pod

=head1 NAME

Clustericious::Config - Configuration files for Clustericious nodes.

=head1 VERSION

version 0.30

=head1 SYNOPSIS

In your ~/etc/MyApp.conf file:

 ---
 % extends_config 'global';
 % extends_config 'hypnotoad', url => 'http://localhost:9999', app => 'MyApp';

 url : http://localhost:9999
 start_mode : hypnotoad
 hypnotoad :
    - heartbeat_timeout : 500
 
 arbitrary_key: value

In your ~/etc/globa.conf file:

 ---
 somevar : somevalue

In your ~/etc/hypnotoad.conf:

 listen :
    - <%= $url %>
 # home uses File::HomeDir to find the calling users'
 # home directory
 pid_file : <%= home %>/<%= $app %>/hypnotoad.pid
 env :
    MOJO_HOME : <%= home %>/<%= $app %>

From a L<Clustericious::App>:

 package MyApp;
 
 use Mojo::Base qw( Clustericious::App );
 
 package MyApp::Routes;
 
 use Clustericious::RouteBuilder;
 
 get '/' => sub {
   my $c = shift;
   my $config = $c; # $config isa Clustericious::Config
   
   # returns the value if it is defined, foo otherwise
   my $value = $config->arbitrary_key(default => 'foo');
 };

From a script:

 use Clustericious::Config;
 
 my $c = Clustericious::Config->new("MyApp");
 my $c = Clustericious::Config->new( \$config_string );
 my $c = Clustericious::Config->new( \%config_data_structure );

 print $c->url;
 print $c->{url};

 print $c->hypnotoad->listen;
 print $c->hypnotoad->{listen};
 my %hash = $c->hypnotoad;
 my @ary  = $c->hypnotoad;

 # Supply a default value for a missing configuration parameter :
 $c->url(default => "http://localhost:9999");
 print $c->this_param_is_missing(default => "something_else");

 # Dump out the entire config as yaml
 print $c->dump_as_yaml;

=head1 DESCRIPTION

Clustericious::Config reads configuration files which are Mojo::Template's
of JSON or YAML files.  There should generally be an entry for
'url', which may be used by either a client or a server depending on
how this node in the cluster is being used.

After rendering the template and parsing the JSON, the resulting
object may be called using method calls or treated as hashes.

Config files are looked for in the following places (in order, where
"MyApp" is the name of the app) :

    $CLUSTERICIOUS_CONF_DIR/MyApp.conf
    $HOME/etc/MyApp.conf
    /util/etc/MyApp.conf
    /etc/MyApp.conf

The helper "extends_config" may be used to read default settings
from another config file.  The first argument to extends_config is the
basename of the config file.  Additional named arguments may be passed
to that config file and used as variables within that file.  After
reading another file, the hashes are merged (i.e. with Hash::Merge);
so values anywhere inside the data structure may be overridden.

YAML config files must begin with "---", otherwise they are interpreted
as JSON.

This module provides a number of helpers
which can be used to get system details (such as the home directory of
the calling user or to prompt for passwords).  See L<Clustericious::Config::Helpers>
for details.

=head1 CONSTRUCTOR

=head2 new

Create a new Clustericious::Config object.  See the SYNOPSIS for
possible invocations.

=head1 METHODS

=head2 $config-E<gt>dump_as_yaml

Returns a string with the configuration encoded as YAML.

=head2 Clustericious::Config->set_singleton

Cache a config object to be returned by the constructor.  Usage:

 Clustericicious::Config->set_singleton(App => $object);

=head1 CAVEATS

Some filesystems do not support filenames with a colon
(:) character in them, so for apps with a double colon
in them (for example L<Clustericious::HelloWorld>),
a single dash character will be substituted for the name
(for example C<Clustericious-HelloWorld.conf>).

=head1 NOTES

This is a beta release. The API may change without notice.

=head1 SEE ALSO

L<Mojo::Template>, L<Hash::Merge>, L<Clustericious>, L<Clustericious::Client>, L<Clustericious::Config::Helpers>

=head1 AUTHOR

original author: Brian Duggan

current maintainer: Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

